java/src/jmri/jmrit/display/layoutEditor/configurexml/LayoutTurnoutXml.java

```diff
@@ -191,21 +189,32 @@ public void load(Element element, Object o) {
         String name = element.getAttribute("ident").getValue();
         double x = 0.0;
         double y = 0.0;
-        LayoutTurnout.TurnoutType tType = LayoutTurnout.TurnoutType.RH_TURNOUT;
+
         try {
             x = element.getAttribute("xcen").getFloatValue();
             y = element.getAttribute("ycen").getFloatValue();
-            tType = tTypeEnumMap.inputFromAttribute(element.getAttribute("type"));
-        } catch (org.jdom2.DataConversionException e) {
-            log.error("failed to convert layoutturnout attribute");
+        }
+        catch (DataConversionException e) {
+            log.error("failed to convert layoutturnout x & y attributes");
+        }
+
+        LayoutTurnout.TurnoutType tType = null;
+        Attribute a = element.getAttribute("type");
+        if (a != null) {
+            tType = LayoutTurnout.TurnoutType.getValue(a.getValue());
```

Comment:
The ENumIO method (1) lets this read existing numeric files for now and (2) lets us change to names later by changing just that one class.  That change will require changes to schema and documentation, so it makes sense to do it all at once rather than over and over ask individual sets of constants are converted to an enum.

Reply:
My code also supports the numeric files and (enum name) strings. This could be written as a enum class and then subclassed with the specific values.

You obviously trust the ordinals more that I do… The problem with making something idiot proof is god keeps making better idiots. I'm more of a "belt and suspenders" kind of guy but I'm not married to my method. I'm ok with doing it your way.

In any case after your PR is merged I'll back out my specific changes in order to fix all the conflicts your PR will have with this one.

Reply:
If you care how I did it check out the "LayoutTurnout.getValue(,,,)" method. That's the code I use to initially match against the enum name strings and if that fails fall back on matching against the xmlValue. The numeric testing could have just as easily been written to match against the ordinals.

Reply:
Also note that you can't just store the enum as-is and depend on .toString() to save it verbatim. The .toString() method could be modified/overridden to include additional (debugging?) information; this would make the saved string not match the actual enum names. It's saver to store the enum.name() string.;

Reply:
Also, I question the need for a redundant copy of the enum values(). If the code used the local copy more than once it might be worth it but with the code being called over and over again for each LayoutTurnout element it's just unnecessary overhead.

Reply:
The purpose of enums is to be blind types; every time you expose their internals in normal operation, they lose value.  That's why the ones I added had no methods to expose a number or string name (they do have some logic and comparison methods)

For I/O, I agree we do need to have a mapping from enum value <-> numbers now for backward file compatibility (later, not). If you don't trust the ordinal number, that's OK, just put that mapping explicitly in the EnumIO instance.

Reply:
(I don't know what you mean by redundant copies of enum values)

Reply:
Not enum values… enum values(); EnumIO makes a copy of the array of all the valid values:
`EnumIO<LayoutTurnout.LinkType> linkEnumMap = new EnumIO<>(LayoutTurnout.LinkType.class); `
What is the value added in doing this?

Reply:
> just put that mapping explicitly in the EnumIO instance.
Which is what I do… but your solution ignores that mapping and just uses the (fragile) ordinal.

Reply:
I'm thinking that the correct solution would be to create an abstract base class enum with a combination of your code and mine… Enum sub-classes without hard coded xml values would "just work" with a getValue method based on ordinals; sub-classes with hard coded xml values would have to override the getValue method to match against those values instead. The base class would have your getAttribute method modified to use the getValue method instead of being hard coded for the ordinal.
Just my thoughts… YOMV.

Reply:
Correction: Your enum doesn't copy the values (my bad: I read "map" and my brain thought "HashMap" and jumped to the wrong conclusion.)

But… outputFromEnum should return e.name(); not e.ordinal().

inputFromString should be in the enum implementation and use ordinals for enums without xml values and use the xml values otherwise.

Reply:
I agree that exposing the ordinals isn't great, and that it should be replaced _everywhere_ with using the enum value itself.

That includes reading and writing the enum values, but JMRI can't write out and read back strings without modifying the schema.  We'll get to that, but not now.

I understand that improving a long LayoutEditor history is incremental, and we can't do everything all at once.  Using enums to the extent we can is a good thing.  Eventually, it would be good to reduce the use of their ordinals even more.   I've added some tests to make sure the ordinals are stable (i.e the test will prevent somebody from changing the order) in the enums I added to try to help with that in the meantime.

Reply:
My PR modifies the appropriate schema's to accept a string instead of an int. note: the int's were being stored as strings (with double quotes, example: type="7") so this is backwards compatible.
I agree the enum tests are a good thing (I added them to mine) but they don't prevent a developer that doesn't know better from just "fixing" the tests when they fail.